File: backend/src/main/java/com/ingilizce/calismaapp/service/ChatbotService.java
package com.ingilizce.calismaapp.service;

import dev.langchain4j.service.SystemMessage;
import dev.langchain4j.service.UserMessage;
import dev.langchain4j.service.spring.AiService;

@AiService
public interface ChatbotService {

    /**
     * Cümle üretme servisi - UNIVERSAL MODE
     * Kelimenin farklı anlamlarını (polisemi) tarar ve doğru bağlamda çeviri yapar.
     */
    @SystemMessage("""
        ROLE: Expert English Lexicographer and Translator.
        
        TASK:
        Generate 5 distinct English sentences that demonstrate the usage of the user's target word/phrase.
        
        CRITICAL INSTRUCTIONS:
        1. **Variety is Key:** If the word has multiple meanings (polysemy), generate sentences that cover DIFFERENT meanings.
           - Example for 'run': One sentence for "physical running", one for "managing a business", one for "machine operating".
        2. **Contextual Translation:** The 'turkishTranslation' field must be the exact equivalent of the target word IN THAT SPECIFIC SENTENCE context. Do NOT translate the whole sentence, just the target word's meaning in that context (1-3 words).
        3. **Full Translation:** The 'turkishFullTranslation' field must be the complete, natural Turkish translation of the entire English sentence.
        4. **Difficulty Level:** Adjust sentence complexity based on the specified CEFR level (A1=very simple, A2=simple, B1=intermediate, B2=upper-intermediate, C1=advanced, C2=very advanced).
        5. **Sentence Length:** Adjust sentence length based on the specified length (short=5-8 words, medium=9-15 words, long=16+ words).
        6. **No Hallucinations:** Do not invent words. Use standard Turkish dictionary meanings.
        
        OUTPUT FORMAT (CRITICAL - MUST BE A JSON ARRAY, NOT AN OBJECT):
        Return ONLY a JSON array, starting with [ and ending with ]. Do NOT wrap it in an object.
        [
          {"englishSentence": "...", "turkishTranslation": "...", "turkishFullTranslation": "..."},
          {"englishSentence": "...", "turkishTranslation": "...", "turkishFullTranslation": "..."},
          ...
        ]
        
        IMPORTANT: 
        - Return ONLY the JSON array, nothing else
        - Do NOT add any text before or after the array
        - Do NOT wrap the array in an object like {"sentences": [...]}
        - Start directly with [ and end with ]
        
        ONE-SHOT EXAMPLES (Study these carefully):
        
        Input: "book" (Level: B1, Length: medium)
        Output:
        [
          {"englishSentence": "I am reading a good book.", "turkishTranslation": "kitap", "turkishFullTranslation": "İyi bir kitap okuyorum."},
          {"englishSentence": "I need to book a hotel room.", "turkishTranslation": "rezervasyon yapmak", "turkishFullTranslation": "Bir otel odası rezervasyonu yapmam gerekiyor."},
          {"englishSentence": "The police booked him for speeding.", "turkishTranslation": "ceza yazmak", "turkishFullTranslation": "Polis ona hız sınırını aştığı için ceza yazdı."},
          {"englishSentence": "She wrote a book about cats.", "turkishTranslation": "kitap", "turkishFullTranslation": "Kediler hakkında bir kitap yazdı."},
          {"englishSentence": "The flight is fully booked.", "turkishTranslation": "dolu", "turkishFullTranslation": "Uçuş tamamen dolu."}
        ]
        """)
    @UserMessage("Target word: '{{it}}'. Generate sentences in pure JSON.")
    String generateSentences(String message);

    /**
     * Çeviri kontrolü servisi
     */
    @SystemMessage("""
        ROLE: You are a supportive and encouraging English-Turkish translation checker.
        
        TASK:
        1. Evaluate the user's Turkish translation for the given English sentence.
        2. Be GENEROUS and SUPPORTIVE - if the translation is mostly correct or conveys the meaning well, mark it as CORRECT.
        3. Only mark as INCORRECT if there are significant meaning errors or major grammar mistakes.
        
        CRITICAL RULES:
        - Focus on MEANING and GRAMMAR, NOT minor spelling mistakes or typos.
        - IGNORE small typos like: missing/extra letters, capitalization errors, punctuation mistakes, or single character errors.
        - If the translation conveys the correct meaning and grammar is mostly correct, mark it as CORRECT.
        - Be LENIENT: Multiple acceptable translations exist. If the user's translation is reasonable and conveys the meaning, it's CORRECT.
        - Only mark as INCORRECT if: meaning is significantly wrong, grammar is fundamentally broken, or there are multiple major errors.
        - When CORRECT: Provide positive, encouraging feedback in Turkish. You can suggest minor improvements as "tips" but still mark as correct.
        - When INCORRECT: Provide the correct translation and explain the mistake clearly and constructively.
        - IMPORTANT: If the user's translation is similar to a standard translation (even if worded slightly differently), mark it as CORRECT and provide encouraging feedback with optional suggestions.
        - Provide clear, concise, supportive feedback in Turkish.
        - Return ONLY a JSON object with this exact format:
        {
          "isCorrect": true or false,
          "correctTranslation": "correct Turkish translation here (only if isCorrect is false, or as a reference if correct)",
          "feedback": "encouraging explanation in Turkish (positive feedback if correct, constructive error explanation if incorrect)"
        }
        - Do not add any text before or after the JSON.
        """)
    @UserMessage("{{it}}")
    String checkTranslation(String message);

    /**
     * İngilizce sohbet pratiği servisi - Buddy Mode
     */
    @SystemMessage("""
        You are Owen, a friendly English chat buddy. NOT a teacher. Just a friend chatting.
        
        STRICT RULES:
        1. MAX 8-10 words per sentence. Break long thoughts into short sentences.
        2. ALWAYS start with a filler: "Alright...", "Nice!", "Hmm...", "Well...", "Okay...", "Oh!", "Cool!"
        3. ALWAYS end with a question to keep conversation going.
        4. Use contractions: I'm, you're, don't, can't, won't, let's, that's.
        5. NO teaching. NO grammar explanations. Just chat like a buddy.
        6. If user makes a mistake, don't correct formally. Just naturally use the correct form.
        
        RESPONSE FORMAT:
        [Filler] + [1-2 short sentences] + [Question]
        
        EXAMPLES:
        User: "I go to school yesterday"
        You: "Nice! So you went to school. What did you do there?"
        
        User: "Hello"
        You: "Hey! Good to hear you. How's your day going?"
        
        User: "I am fine"
        You: "Awesome! Glad to hear that. What are you up to today?"
        
        NEVER:
        - Write more than 3 short sentences
        - Give grammar lessons
        - Use formal language
        - Skip the filler at the start
        - Skip the question at the end
        """)
    @UserMessage("{{it}}")
    String chat(String message);

    /**
     * IELTS/TOEFL Speaking test soruları üretme servisi
     */
    @SystemMessage("""
        ROLE: Expert IELTS/TOEFL Speaking Test Examiner
        
        TASK:
        Generate authentic IELTS/TOEFL Speaking test questions based on the test type and part.
        
        FORMAT:
        - IELTS Part 1: Personal questions (hometown, work, studies, hobbies) - 3-4 questions
        - IELTS Part 2: Cue card with topic (describe, explain, discuss) - 1 question with 3-4 sub-points
        - IELTS Part 3: Abstract discussion questions related to Part 2 topic - 3-4 questions
        - TOEFL Task 1: Independent speaking (personal opinion) - 1 question
        - TOEFL Task 2-4: Integrated speaking (read/listen/speak) - 1 question with context
        
        Return ONLY a JSON object with this format:
        {
          "questions": ["question1", "question2", ...],
          "instructions": "specific instructions for this part",
          "timeLimit": seconds,
          "preparationTime": seconds (if applicable)
        }
        """)
    @UserMessage("Generate {{testType}} Speaking test questions for {{part}}. Return ONLY JSON.")
    String generateSpeakingTestQuestions(String message);

    /**
     * IELTS/TOEFL Speaking test puanlama servisi
     */
    @SystemMessage("""
        ROLE: Expert IELTS/TOEFL Speaking Test Examiner
        
        TASK:
        Evaluate the candidate's speaking performance and provide detailed scores and feedback.
        
        IELTS SCORING (0-9 for each criterion, then average):
        1. Fluency and Coherence (0-9): Smoothness, natural flow, logical organization
        2. Lexical Resource (0-9): Vocabulary range, accuracy, appropriateness
        3. Grammatical Range and Accuracy (0-9): Grammar variety, complexity, errors
        4. Pronunciation (0-9): Clarity, intonation, stress, accent (not native accent requirement)
        
        TOEFL SCORING (0-30 total):
        1. Delivery (0-10): Clear pronunciation, natural pace, intonation
        2. Language Use (0-10): Grammar, vocabulary accuracy and range
        3. Topic Development (0-10): Ideas, organization, completeness
        
        CRITICAL RULES:
        - Be FAIR and CONSISTENT with official IELTS/TOEFL standards
        - Provide specific examples from the candidate's response
        - Give constructive feedback for improvement
        - Score realistically (not too harsh, not too lenient)
        - Consider that this is practice, so be encouraging but accurate
        
        Return ONLY a JSON object with this format:
        {
          "overallScore": number (IELTS: 0-9, TOEFL: 0-30),
          "criteria": {
            "fluency": number (IELTS only),
            "lexicalResource": number (IELTS only),
            "grammar": number (IELTS only),
            "pronunciation": number (IELTS only),
            "delivery": number (TOEFL only),
            "languageUse": number (TOEFL only),
            "topicDevelopment": number (TOEFL only)
          },
          "feedback": "detailed feedback in Turkish",
          "strengths": ["strength1", "strength2", ...],
          "improvements": ["improvement1", "improvement2", ...]
        }
        """)
    @UserMessage("Evaluate this {{testType}} Speaking test response. Question: {{question}}. Candidate's response: {{response}}. Return ONLY JSON.")
    String evaluateSpeakingTest(String message);
}

File: flutter_vocabmaster/lib/services/chatbot_service.dart
import 'dart:convert';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:http/http.dart' as http;
import '../config/app_config.dart';

/// Chatbot servisi - Direkt Groq API kullanır (backend bağımsız)
/// Cümle üretimi, çeviri kontrolü ve chatbot işlemleri için kullanılır.
class ChatbotService {
  static const String _groqUrl = 'https://api.groq.com/openai/v1/chat/completions';
  static String get _apiKey => dotenv.env['GROQ_API_KEY'] ?? '';

  /// Kelime için pratik cümleleri üretir - DİREKT GROQ API
  Future<Map<String, dynamic>> generateSentences({
    required String word,
    List<String> levels = const ['B1'],
    List<String> lengths = const ['medium'],
    bool checkGrammar = false,
  }) async {
    if (_apiKey.isEmpty) {
      throw Exception('GROQ_API_KEY bulunamadı. .env dosyasını kontrol edin.');
    }

    // Random seed for variety
    final randomSeed = DateTime.now().millisecondsSinceEpoch;

    final prompt = '''
Generate 3 COMPLETELY NEW and UNIQUE English sentences using the word/phrase "$word".
Target levels: ${levels.join(', ')}
Target lengths: ${lengths.join(', ')} (short=5-8 words, medium=9-15 words, long=16+ words)

IMPORTANT: 
- Generate DIFFERENT sentences each time - be creative and varied!
- Use different contexts, scenarios, and sentence structures.
- Random seed for this request: $randomSeed

For each sentence, provide:
1. The English sentence
2. A natural, fluent Turkish translation (NOT word-for-word)

Return ONLY a valid JSON object in this exact format:
{
  "sentences": ["English sentence 1", "English sentence 2", "English sentence 3"],
  "translations": ["Türkçe çeviri 1", "Türkçe çeviri 2", "Türkçe çeviri 3"]
}
''';

    try {
      final response = await http.post(
        Uri.parse(_groqUrl),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $_apiKey',
        },
        body: jsonEncode({
          "model": "llama-3.3-70b-versatile",
          "messages": [
            {"role": "system", "content": "You are an expert English-Turkish translator. Generate NEW, CREATIVE, and UNIQUE sentences each time. Never repeat the same sentences. Return ONLY valid JSON."},
            {"role": "user", "content": prompt}
          ],
          "temperature": 1.0,
          "top_p": 0.95,
          "response_format": {"type": "json_object"}
        }),
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final data = jsonDecode(utf8.decode(response.bodyBytes));
        final content = data['choices'][0]['message']['content'];
        return jsonDecode(content);
      } else {
        print('Groq generateSentences error: ${response.statusCode} - ${response.body}');
        throw Exception('Cümleler üretilemedi: ${response.statusCode}');
      }
    } catch (e) {
      print('ChatbotService.generateSentences error: $e');
      rethrow;
    }
  }

  /// Kullanıcının çevirisini kontrol eder - DİREKT GROQ API
  Future<Map<String, dynamic>> checkTranslation({
    required String originalSentence,
    required String userTranslation,
    required String direction, // 'EN_TO_TR' or 'TR_TO_EN'
    String? referenceSentence,
  }) async {
    if (_apiKey.isEmpty) {
      throw Exception('GROQ_API_KEY bulunamadı.');
    }

    String prompt;
    if (direction == 'EN_TO_TR') {
      prompt = '''
Check this English to Turkish translation:

English: "$originalSentence"
User's Turkish translation: "$userTranslation"

Be GENEROUS - if the meaning is correct and grammar is mostly right, mark as correct.
Ignore minor typos or spelling errors.

Return ONLY a JSON object:
{
  "isCorrect": true or false,
  "correctTranslation": "doğru Türkçe çeviri",
  "feedback": "Türkçe açıklama - teşvik edici olun"
}
''';
    } else {
      prompt = '''
Check this Turkish to English translation:

Turkish: "$originalSentence"
User's English translation: "$userTranslation"
${referenceSentence != null ? 'Reference: "$referenceSentence"' : ''}

Be GENEROUS - if the meaning is correct and grammar is mostly right, mark as correct.
Ignore minor typos.

Return ONLY a JSON object:
{
  "isCorrect": true or false,
  "correctTranslation": "correct English translation",
  "feedback": "explanation in Turkish"
}
''';
    }

    try {
      final response = await http.post(
        Uri.parse(_groqUrl),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $_apiKey',
        },
        body: jsonEncode({
          "model": "llama-3.3-70b-versatile",
          "messages": [
            {"role": "system", "content": "You are a supportive translation checker. Return ONLY valid JSON."},
            {"role": "user", "content": prompt}
          ],
          "temperature": 0.3,
          "response_format": {"type": "json_object"}
        }),
      ).timeout(const Duration(seconds: 20));

      if (response.statusCode == 200) {
        final data = jsonDecode(utf8.decode(response.bodyBytes));
        final content = data['choices'][0]['message']['content'];
        return jsonDecode(content);
      } else {
        throw Exception('Çeviri kontrol edilemedi: ${response.statusCode}');
      }
    } catch (e) {
      print('ChatbotService.checkTranslation error: $e');
      rethrow;
    }
  }

  /// AI Bot ile sohbet - DİREKT GROQ API
  Future<String> chat(String message) async {
    if (_apiKey.isEmpty) {
      throw Exception('GROQ_API_KEY bulunamadı.');
    }

    final systemPrompt = '''
You are Owen, a friendly English chat buddy. NOT a teacher. Just a friend chatting.

RULES:
1. MAX 10-12 words per sentence. Keep it short.
2. Start with a filler: "Alright...", "Nice!", "Hmm...", "Well...", "Okay...", "Oh!", "Cool!"
3. End with a question to keep conversation going.
4. Use contractions: I'm, you're, don't, can't, won't, let's.
5. NO teaching or grammar explanations. Just chat naturally.
6. If user makes a mistake, naturally use the correct form without correcting.

Format: [Filler] + [1-2 short sentences] + [Question]
''';

    try {
      final response = await http.post(
        Uri.parse(_groqUrl),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $_apiKey',
        },
        body: jsonEncode({
          "model": "llama-3.3-70b-versatile",
          "messages": [
            {"role": "system", "content": systemPrompt},
            {"role": "user", "content": message}
          ],
          "temperature": 0.8,
          "max_tokens": 150
        }),
      ).timeout(const Duration(seconds: 20));

      if (response.statusCode == 200) {
        final data = jsonDecode(utf8.decode(response.bodyBytes));
        return data['choices'][0]['message']['content'];
      } else {
        throw Exception('Chat failed: ${response.statusCode}');
      }
    } catch (e) {
      print('ChatbotService.chat error: $e');
      rethrow;
    }
  }

  /// Kelimeyi bugüne kaydet - BACKEND API (kelime kaydetme işlemi için)
  Future<Map<String, dynamic>> saveWordToToday({
    required String englishWord,
    List<String> meanings = const [],
    List<String> sentences = const [],
  }) async {
    final baseUrl = await AppConfig.apiBaseUrl;
    
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/chatbot/save-to-today'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'englishWord': englishWord,
          'meanings': meanings,
          'sentences': sentences,
        }),
      ).timeout(const Duration(seconds: 15));
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('Kelime kaydedilemedi: ${response.statusCode}');
      }
    } catch (e) {
      print('ChatbotService.saveWordToToday error: $e');
      rethrow;
    }
  }

  /// IELTS/TOEFL Speaking test soruları oluştur - DİREKT GROQ API
  Future<Map<String, dynamic>> generateSpeakingTestQuestions({
    required String testType, // 'IELTS' or 'TOEFL'
    required String part,
  }) async {
    if (_apiKey.isEmpty) {
      throw Exception('GROQ_API_KEY bulunamadı.');
    }

    final prompt = '''
Generate authentic $testType Speaking test questions for $part.

For IELTS:
- Part 1: Personal questions (3-4 questions about everyday topics)
- Part 2: Cue card topic with sub-points
- Part 3: Abstract discussion questions (3-4 questions)

For TOEFL:
- Task 1: Independent speaking (personal opinion question)
- Task 2-4: Integrated speaking questions

Return ONLY a JSON object:
{
  "question": "The main question to ask",
  "questions": ["question1", "question2", ...],
  "instructions": "specific instructions for this part",
  "timeLimit": 60
}
''';

    try {
      final response = await http.post(
        Uri.parse(_groqUrl),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $_apiKey',
        },
        body: jsonEncode({
          "model": "llama-3.3-70b-versatile",
          "messages": [
            {"role": "system", "content": "You are an expert $testType examiner. Return ONLY valid JSON."},
            {"role": "user", "content": prompt}
          ],
          "temperature": 0.7,
          "response_format": {"type": "json_object"}
        }),
      ).timeout(const Duration(seconds: 30));

      if (response.statusCode == 200) {
        final data = jsonDecode(utf8.decode(response.bodyBytes));
        final content = data['choices'][0]['message']['content'];
        return jsonDecode(content);
      } else {
        throw Exception('Sorular oluşturulamadı: ${response.statusCode}');
      }
    } catch (e) {
      print('ChatbotService.generateSpeakingTestQuestions error: $e');
      rethrow;
    }
  }

  /// Speaking test cevabını değerlendir - DİREKT GROQ API
  Future<Map<String, dynamic>> evaluateSpeakingTest({
    required String testType,
    required String question,
    required String response,
  }) async {
    if (_apiKey.isEmpty) {
      throw Exception('GROQ_API_KEY bulunamadı.');
    }

    final prompt = '''
Evaluate this $testType Speaking test response:

Question: "$question"
Candidate's Response: "$response"

For IELTS, score 0-9 on: Fluency, Lexical Resource, Grammar, Pronunciation
For TOEFL, score 0-30 total on: Delivery, Language Use, Topic Development

Be fair and encouraging. Provide constructive feedback.

Return ONLY a JSON object:
{
  "score": number (IELTS: 0-9, TOEFL: 0-30),
  "band": "score as string",
  "feedback": "detailed feedback in Turkish",
  "suggestions": "improvement suggestions in Turkish"
}
''';

    try {
      final httpResponse = await http.post(
        Uri.parse(_groqUrl),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $_apiKey',
        },
        body: jsonEncode({
          "model": "llama-3.3-70b-versatile",
          "messages": [
            {"role": "system", "content": "You are an expert $testType examiner. Return ONLY valid JSON."},
            {"role": "user", "content": prompt}
          ],
          "temperature": 0.5,
          "response_format": {"type": "json_object"}
        }),
      ).timeout(const Duration(seconds: 30));

      if (httpResponse.statusCode == 200) {
        final data = jsonDecode(utf8.decode(httpResponse.bodyBytes));
        final content = data['choices'][0]['message']['content'];
        return jsonDecode(content);
      } else {
        throw Exception('Değerlendirme başarısız: ${httpResponse.statusCode}');
      }
    } catch (e) {
      print('ChatbotService.evaluateSpeakingTest error: $e');
      rethrow;
    }
  }
}
